### ISSUE TITLE
Stratum source mismatch in computing total biomass density

### ISSUE THREAD

**Issue by leewujung** [OSOceanAcoustics/echopop#38]
From meeting on 2022/11/17:
- The current NASC file contains a column called `Stratum` and a column called `Assigned_haul`
- Most rows have `Assigned_haul=0`, and for these rows a `Stratum` number is assigned outside of EchoPro
- Previously @b-reyes and @emiliom traced the code and various sources and determined that this `Stratum` number refers the "Base KS" stratum (rather than INPFC stratum)

This creates a problem in the code below https://github.com/uw-echospace/EchoPro/blob/689ba79220353e5e994e533b6b816d6071da5a0a/EchoPro/compute_biomass_density/compute_biomass_density.py#L800-L802
where the referenced stratum number in `self.nasc_df.Stratum` can mismatch the stratum number user intended to use from ` self.strata_sig_b`. Specifically:
- if user selects to use "Base KS" stratum, everything will be fine
- if user selects to use "INPFC" stratum, the stratum mismatch happens

@b-reyes : the above is what I understand so far, please edit directly if you see any errors!

**Comment by leewujung** [comment]
@brandynlucca : We've been talking mostly about the "Base KS" stratification so far. Correct me if I am wrong, but I don't think we've done anything with INPFC strata yet. Seems like to get everything straightened out, we'll have to have separate columns `stratum_KS` and `stratum_INPFC`. Good thing is `stratum_INPFC` just based on latitude, so would be easy to construct that column. 

I am adding this under v0.4.0 refactoring right now, but we can also deal with it after the bootstrapping stuff is overhauled.

**Comment by brandynlucca** [comment]
> @brandynlucca : We've been talking mostly about the "Base KS" stratification so far. Correct me if I am wrong, but I don't think we've done anything with INPFC strata yet. Seems like to get everything straightened out, we'll have to have separate columns `stratum_KS` and `stratum_INPFC`. Good thing is `stratum_INPFC` just based on latitude, so would be easy to construct that column.

That is correct. The `stratum_INPFC` term doesn't come up until it is assigned to each transect within `run_jolly_hampton` in the current implementation that I have not refactored yet via `get_transect_strata_info_no_kriging` and `get_transect_strata_info_kriging`. As far as I can tell, this is currently hard-coded as the default argument for `lat_infpc` within `run_cv_analysis` such that:

```python
def run_cv_analysis(
        self,
        lat_inpfc: Tuple[float] = (np.NINF, 36, 40.5, 43.000, 45.7667, 48.5, 55.0000),
        kriged_data=False,
        seed=None,
    ) -> float:
```

This replicates what is contained within the `INPFC` sheet found in (at least for this dataset) `Stratification_geographic_Lat_2019_final.xlsx` (which is read in to create `geostrata_df` from the `stratification1` in the refactored implementation). Instead of continuing to hard-code the values as a default argument, I don't think it would be particularly arduous to read in the `INFPC` sheet alongside `stratification1`. That way when `EchoPro.Survey` is initialized, an additional code block in `load_survey_data` that merges these strata into `nasc_df` similar to how `self.biology['haul_to_transect_df']` is used as a cypher (of sorts) to match the `haul_num`, `transect_num`, and `stratum_num` variables found in all of the biological datasets, i.e.:

```python
        ### Merge haul numbers and regional indices across biological variables
        # Also add strata values/indices here alongside transect numbers 
        # -- Step 1: Consolidate information linking haul-transect-stratum
        self.biology['haul_to_transect_df'] = (
            self.biology['haul_to_transect_df']
            .merge(self.spatial['strata_df'], on ='haul_num' , how = 'outer' )
        )
        # -- Step 2: Distribute this information to the other biological variables
        # ---- Specimen 
        self.biology['specimen_df'] = (
            self.biology['specimen_df']
            .merge( self.biology['haul_to_transect_df'] , on = ['haul_num' , 'region' ] ) 
        )
 ```

Given that the sheet has both the `Latitude (upper limit)` _and_ the `haul start` and `haul end`, I think either latitude or haul could be used to index (the former for `nasc_df`, the latter two for `haul_num` indexed dataframes like `specimen_df`). I have already been playing around with this a little bit, so I think something _like_ the following example would be workable that would resemble how other similar operations have been refactored. I am including this just as a timestamped reminder in case my understanding of how these strata are incorporated is incorrect.

_Preparing the example and approximating how the data structure would be stored within the `EchoPro.Survey` object._

```python
### Reference NASC data
nasc_df = self.acoustics['nasc']['nasc_df']

### Generate example INPFC DataFrame
INPFC_df = pd.DataFrame(
    {
        'INPFC_stratum': np.array( [ 1 , 2 , 3 , 4 , 5 , 6 ] , dtype = int ) ,
        'latitude_limits': np.array( [ 36.0 , 40.5 , 43.0 , 45.7667 , 48.5 , 55.0 ] ) ,
        'haul_start': np.array( [ 1 , 8 , 35 , 47 , 62 , 201 ] , dtype = int ) ,
        'haul_end': np.array( [ 7 , 33 , 46 , 61 , 73 , 222 ] , dtype = int )
    }
)

# Pre- and append end-values to complete intervals
latitude_bins = np.concatenate( [ [ -90 ] , INPFC_df.latitude_limits , [ 90 ] ] )
haul_bins = np.sort( np.concatenate( [ INPFC_df[ 'haul_start' ] , INPFC_df[ 'haul_end' ] ] ) )

# Assign intervals to INPFC_df
INPFC_binned_df = (
    INPFC_df
    .pipe( lambda df: df
          .assign( latitude_stratum = lambda x: pd.cut( x[ 'latitude_limits' ] , latitude_bins ) , 
                   haul_stratum = lambda x: pd.cut( ( x[ 'haul_start' ] + x[ 'haul_end' ] ) / 2 , haul_bins ) ) )
)
```

```python
>>> INPFC_binned_df
   INPFC_stratum  latitude_limits  haul_start  haul_end latitude_stratum haul_stratum
0              1          36.0000           1         7    (-90.0, 36.0]       (1, 7]
1              2          40.5000           8        33     (36.0, 40.5]      (8, 33]
2              3          43.0000          35        46     (40.5, 43.0]     (35, 46]
3              4          45.7667          47        61   (43.0, 45.767]     (47, 61]
4              5          48.5000          62        73   (45.767, 48.5]     (62, 73]
5              6          55.0000         201       222     (48.5, 55.0]   (201, 222]
```

_Merging the INPFC stratum definitions into `nasc_df` using both binned columns/variables_

```python
### Merge INPFC-defined stratum using latitude
nasc_latitude = (
    nasc_df
    .pipe( lambda df: df
           .assign( latitude_stratum = lambda x: pd.cut( x[ 'latitude' ] , latitude_bins ) ) ) 
    .merge( INPFC_binned_df[ [ 'INPFC_stratum' , 'latitude_stratum' ] ] , on = 'latitude_stratum' )
)
```

```python
>>> nasc_latitude
      transect_num  vessel_log_start  vessel_log_end   latitude   longitude  stratum_num  transect_spacing  NASC_no_age1  haul_num  NASC_all_ages latitude_stratum  INPFC_stratum
0                1        744.016009      744.491145  34.397267 -121.143005            1              10.0           0.0         0            0.0    (-90.0, 36.0]              1
1                1        744.500276      744.995605  34.397391 -121.133196            1              10.0           0.0         0            0.0    (-90.0, 36.0]              1
2                1        745.004125      745.499447  34.397435 -121.123057            1              10.0           0.0         0            0.0    (-90.0, 36.0]              1
3                1        745.508199      745.994306  34.397394 -121.112871            1              10.0           0.0         0            0.0    (-90.0, 36.0]              1
4                1        746.003214      746.495701  34.397437 -121.102888            1              10.0           0.0         0            0.0    (-90.0, 36.0]              1
...            ...               ...             ...        ...         ...          ...               ...           ...       ...            ...              ...            ...
9273           145       3073.001555     3073.499465  51.897250 -131.025358            1              10.0           0.0         0            0.0     (48.5, 55.0]              6
9274           145       3073.501846     3073.998088  51.897212 -131.011859            1              10.0           0.0         0            0.0     (48.5, 55.0]              6
9275           145       3074.000343     3074.498887  51.896882 -130.998452            1              10.0           0.0         0            0.0     (48.5, 55.0]              6
9276           145       3074.501062     3074.998484  51.897065 -130.984977            1              10.0           0.0         0            0.0     (48.5, 55.0]              6
9277           145       3075.000658     3075.263539  51.896152 -130.971639            1              10.0           0.0         0            0.0     (48.5, 55.0]              6

[9278 rows x 12 columns]
```

**Comment by leewujung** [comment]
@brandynlucca : Thanks for the detailed response, it's comprehensive, I think what you proposed would work well as part of `__init__` (or a function called by it).

The only thing I would suggest is to rename `latitude_stratum` and `haul_stratum` to `INPFC_stratum_latitude_interval` and `INPFC_strate_haul_interval`: I first saw these in the `nasc_latitude` printout and was confused (I have a weird habit of reading from the bottom up...). I thought the suggested names are nice because the names would make it clear that these columns contain some kind of intervals associated with INPFC.

I agree with initiating `INPFC_df` in `Survey.__init__` using `Stratification_geographic_Lat_2019_final.xlsx`.
A note related to this: I know that Emilio manually changed the column `Latitude (upper limit)` to `northlimit_latitude`, and if I understand correctly, Alicia will change her code so that when these files are generated from the FEAT database in the future that column will be named as `northlimit_latitude`. We can/should double check with Alicia about that.

Another thing that I wonder about is: Wouldn't `Latitude (upper limit)` already sets `haul start` and `haul end`? since any haul would have to be done on a transect, which has a latitude attached to it. I think you may be alluding to this when you italicized the _and_ in below:
> Given that the sheet has both the `Latitude (upper limit)` _and_ the `haul start` and `haul end`, I think either latitude or haul could be used to index (the former for `nasc_df`, the latter two for `haul_num` indexed dataframes like `specimen_df`).

Perhaps it should be that in `Survey.__init__`, `haul start` and `haul end` are set based on `Latitude (upper limit)`. Or, perhaps there should be a check that makes sure that the `haul start` and `haul end` sliced out based on `Latitude (upper limit)` actually correspond to those in the spreadsheet?

**Comment by leewujung** [comment]
This is still TBD from meeting today between @brandynlucca and @leewujung . @brandynlucca will trace the matlab code again to figure out how to resolve this.

It seems that there are 3 possibilities between the stratification for transect analysis and stratification for kriged CV analysis:
- **INPFC** for transect analysis (INPFC strata based on latitude) --> **INPFC** for kriged CV analysis
- **KS-based** for transect analysis (KS strata based on similarity between hauls) --> **geostrata** (geostrata based on latitude derived from KS strata) for kriged CV analysis
- **geostrata** for transect analysis (geostrata based on latitude derived from KS strata) --> **geostrata** (geostrata based on latitude derived from KS strata) for kriged CV analysis

**Comment by brandynlucca** [comment]
This relates to three other reported/tracking issues (#178 and #181, which are all tracked via #228). Georeferenced data can currently be assigned a stratum via 1) the specific haul-stratum pairing (KS) (`strata_df`), 2) latitudinal extents/ranges of different KS strata (`geostrata_df`), and 3) latitudinal extents/regions of INPFC-defined strata (`inpfc_strata_df`). When the key `sigma_bs` is populated, it includes the mean `sigma_bs` per stratum, which depends on the user-defined stratum selection. So when `stratum = 'ks'` (the default case), `sigma_bs` from the KS-defined strata are mapped to the transects that share the same `stratum_num` value determined by their respective latitudes (from `geostrata_df`). When `stratum = 'inpfc'`, the mean `sigma_bs` for each INPFC-defined stratum is mapped to georeferenced values within the same INPFC stratum.

This can be closed once PR #241 is merged.
